content/blog/2019-12-18-ml-based-phases-ordering-with-llvm/index.md

```diff
@@ -0,0 +1,75 @@
+
+### Machine Learning Model
+
+We built a simple linear regression model to predict how much one optimization pass will improve over the performance of the current program, based on a set of performance counters we selected from ones provided by [likwid-perfctr](https://github.com/RRZE-HPC/likwid). During inference time, we will run the model to approximate Hill Climbing algorithm, i.e. greedily select the best pass based on model prediction at each step, until we reach the maximum number of passes limit or there are no more optimizations that will improve the performance.
+
+To collect data points of program performance counters, optimization pass and performace difference, we randomly sampled optimization sequences as in the analysis but with random length. We then collect the runtimes and performance counter values without the last optimization as features. The performance counters we used are:
+
+Clock per Instruction, Branch Misprediction Ratio, Instructions per Branch, % Cycles without Execution, % Cycles Without Execution due to L1D, % Cycles Without Execution due to L2, % Cycles Without Execution due to memory loads, Packed Single Precision MFLOP/S, Packed Double Precision MFLOP/S
+
+We then fit a linear regression model to predict normalized performance improvement. 
+
+## Evaluation
+
+### Hill Climbing baseline
+
+We constructed a optimization sequence of length 100 using Hill Climbing for the program stream cluster:
+
+```
+-loop-distribute -basicaa -instsimplify -aa -lazy-block-freq -domtree -loops -simplifycfg -block-freq -called-value-propagation
+
+-lazy-branch-prob -opt-remark-emitter -basicaa -loop-rotate -correlated-propagation -basicaa -instcombine -simplifycfg -aa -scalar-evolution
+
+-basicaa -simplifycfg -lazy-value-info -aa -transform-warning -basicaa -domtree -aa -basicaa -jump-threading
+
+-loop-accesses -loops -lcssa-verification -domtree -loop-distribute -lazy-block-freq -speculative-execution -instcombine -loops -block-freq
+
+-early-cse -loops -lazy-branch-prob -loop-simplify -domtree -prune-eh -loop-rotate -lcssa-verification -tti -lazy-block-freq
+
+-globalopt -loops -lazy-branch-prob -prune-eh -lcssa -targetlibinfo -aa -loops -mem2reg -constmerge
+
+-basicaa -aa -loop-simplify -loop-unswitch -loop-unroll -basicaa -basicaa -block-freq -tbaa -lazy-branch-prob
+
+-block-freq -lazy-block-freq -basicaa -basicaa -tti -basicaa -basicaa -lazy-branch-prob -loops -opt-remark-emitter
+
+-functionattrs -simplifycfg -loop-simplify -lazy-branch-prob -lazy-block-freq -block-freq -aa -branch-prob -lazy-branch-prob -transform-warning
+
+-aa -domtree -branch-prob -basiccg -scalar-evolution -gvn -lazy-block-freq -basicaa -opt-remark-emitter -loop-accesses
+```
+
+The performance changing is plotted as bellow as passes are added (the box plots are same as in the analysis):
+
+![Hill Climbing order](./wh.png)
+
+As shown in the figure, Hill Climbing gives the almost optimal results.
```

Comment:
I'm not sure what you mean by "optimal results." How do you know what the optimum is?

Reply:
I meant the optimum we found in random search.